runtime/interpreter/interpreter.go

```diff
@@ -2214,8 +2217,182 @@ func (interpreter *Interpreter) writeStored(
 	interpreter.recordStorageMutation()
 }
 
+type FromStringFunctionValue struct {
+	parentType   sema.Type
+	hostFunction *HostFunctionValue
+}
+
+func fromStringDeclaration(ty sema.Type, impl func(*Interpreter, string) Value) FromStringFunctionValue {
+	functionType, _ := sema.FromStringMetadata(ty)
+
+	hostFunctionImpl := NewUnmeteredHostFunctionValue(
+		func(invocation Invocation) Value {
+			argument, ok := invocation.Arguments[0].(*StringValue)
+			if !ok {
+				// expect typechecker to catch a mismatch here
+				panic(errors.NewUnreachableError())
+			}
+			inter := invocation.Interpreter
+			return impl(inter, argument.Str)
+		},
+		functionType,
+	)
+	return FromStringFunctionValue{
+		parentType:   ty,
+		hostFunction: hostFunctionImpl,
+	}
+}
+
+// default implementation for parsing a given unsigned numeric type from a string.
+// the size provided by sizeInBytes is passed to strconv.ParseUint, ensuring that the parsed value fits in the target type.
+// input strings must not begin with a '+' or '-'.
+func fromStringImplUnsigned(bitSize int, convert func(uint64) Value) func(*Interpreter, string) Value {
```

Comment:
`func(*Interpreter, string) Value` is used in multiple places, maybe define a separate type for it.

Also given that these are all returning numeric types, can use the `NumberValue` as the return type. (or should be `OptionalValue`?)